### PR TITLE
Remove WPF dependencies from Workspace and Features test layers

### PR DESCRIPTION
**Pull request by tmat**

### CODE REVIEW

**Comment by tmat** [comment]
@sharwell @jasonmalinowski @dotnet/roslyn-ide PTAL

**Review by genlu — APPROVED** [review]
:shipit:

File: src/Workspaces/CoreTestUtilities/TestExportJoinableTaskContext.cs [hunk 1]
```diff
@@ -37,23 +42,25 @@ public TestExportJoinableTaskContext()
...
+            if (currentContext != null && dispatcherTaskJoiner?.IsDispatcherSynchronizationContext(currentContext) == true)
             {
                 // The current thread is the main thread, and provides a suitable synchronization context
                 mainThread = Thread.CurrentThread;
-                synchronizationContext = SynchronizationContext.Current;
+                synchronizationContext = currentContext;
             }
             else
             {
                 // The current thread is not known to be the main thread; we have no way to know if the
```
Comment (sharwell) [inline src/Workspaces/CoreTestUtilities/TestExportJoinableTaskContext.cs:59]: This needs to have a `Contract.ThrowIfTrue` that verifies the current synchronization context is _not_ a WPF synchronization context.

**Review by sharwell — CHANGES_REQUESTED** [review]
Overall seems fine, but want stronger validation that we don't try to run tests using a WPF dispatcher when `WpfDispatcherTaskJoiner` is missing (see comment in code).

File: src/Workspaces/CoreTestUtilities/TestExportJoinableTaskContext.cs (hunk 1 above)
Comment (tmat) [reply src/Workspaces/CoreTestUtilities/TestExportJoinableTaskContext.cs:59]: Are you saying that if `currentContext` is not null then it must be the WPF dispatcher context?

File: src/Features/CSharpTest/Microsoft.CodeAnalysis.CSharp.Features.UnitTests.csproj [hunk 2]
```diff
@@ -1,10 +1,9 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <!-- Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE file in the project root for more information. -->
-<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
+<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <OutputType>Library</OutputType>
     <TargetFramework>net472</TargetFramework>
```
Comment (dibarbet) [inline src/Features/CSharpTest/Microsoft.CodeAnalysis.CSharp.Features.UnitTests.csproj:6]: Not asking for this to be done in this PR - but this should allow us to target .net core for these tests right?  At some point I want to make that change and have them run on our core CI leg so we have coverage for non-windows platforms (now that this is all running in vscode).

**Review by dibarbet — APPROVED** [review]

File: src/Features/CSharpTest/Microsoft.CodeAnalysis.CSharp.Features.UnitTests.csproj (hunk 2 above)
Comment (tmat) [reply src/Features/CSharpTest/Microsoft.CodeAnalysis.CSharp.Features.UnitTests.csproj:6]: Yes, that'd be the ultimate goal. Unfortunately, it's not simple. Need to disentangle editor dependencies:

https://github.com/dotnet/roslyn/pull/71407

**Review by sharwell — APPROVED** [review]

File: src/Workspaces/CoreTestUtilities/TestExportJoinableTaskContext.cs (hunk 1 above)
Comment (sharwell) [reply src/Workspaces/CoreTestUtilities/TestExportJoinableTaskContext.cs:61]: > Are you saying that if `currentContext` is not null then it must be the WPF dispatcher context?

That's not the case for all applications, but for ours yes.

Comment (tmat) [reply src/Workspaces/CoreTestUtilities/TestExportJoinableTaskContext.cs:62]: Turns out the check fails:

image

**Pull request merged**